specification/dartLangSpec.tex

```diff
@@ -16231,238 +16306,447 @@ \subsection{Imports}
+\commentary{%
+Note that it is a compile-time error for a deferred prefix
+to be used in more than one import,
+which means that the update of the binding of $p$
+does not interfere with other imports.%
 }
 
-\commentary{
-Note that no errors are raised if one hides or shows a name that is not in a namespace.
-}
-\rationale{
-This prevents situations where removing a name from a library would cause breakage of a client library.
+\LMHash{}%
+The effect of a repeated invocation of \code{$p$.loadLibrary()} is as follows:
+\begin{itemize}
+\item
+  If another invocation of \code{$p$.loadLibrary()} has already succeeded,
+  the repeated invocation also succeeds.
+  Otherwise,
+\item
+  If another invocation of \code{$p$.loadLibrary()} has failed:
+  \begin{itemize}
+  \item
+    If the failure is due to a compilation error,
+    the repeated invocation fails for the same reason.
+  \item
+    If the failure is due to other causes,
+    the repeated invocation behaves as if no previous call had been made.
+  \end{itemize}
+\end{itemize}
+
+\commentary{%
+In other words, one can retry a deferred load after a network failure
+or because a file is absent,
+but once one finds some content and loads it, one can no longer reload.
+We do not specify which object the returned future resolves to.%
 }
+\EndCase
 
 \LMHash{}%
-The dart core library \code{dart:core} is implicitly imported into every dart library other than itself via an import clause of the form
+\Case{Semantics of immediate imports}
+When $I_i$ is an immediate import that refers to a URI via the string $s_i$:
+Let $\ell_i$ denote the run-time entity that corresponds to the result of
+compiling the library $L_i$ denoted by $s_i$.
 
-\code{\IMPORT{} `dart:core';}
+\LMHash{}%
+$\ell_i$ is accessible in the current isolate at run time
+due to an implementation specific mechanism,
+and it provides access to the result of compiling functions,
+as well as other resources needed for the execution of code in $L_i$.
 
-unless the importing library explicitly imports \code{dart:core}.
+\LMHash{}%
+Let \NamespaceName{\metavar{import}, i} be the namespace imported from $L_i$.
+The run-time namespace \NamespaceName{i} will then have the following bindings:
 
-\commentary{
-Any import of \code{dart:core}, even if restricted via \SHOW{}, \HIDE{} or \AS{}, preempts the automatic import.
-}
+\begin{itemize}
+\item
+  For every top level function, getter, or setter $m$ named $n$ in
+  \NamespaceName{\metavar{import}, i},
+  a binding from $n$ to the result of compiling $m$,
+  obtained from $\ell_i$.
+\item
+  For every name \id{} in \NamespaceName{\metavar{import}, i}
+  that is bound to a class, mixin, or type alias declaration
+  introducing a type $T$,
+  a binding from \id{} to a compiled getter with return type \code{Type}
+  that, when invoked, returns the type object for $T$.
```

Comment:
This doesn't work correctly for accessing static methods of classes.

Just say that you export the real declaration. It's a static error to use the type as a *type*, so you don't need to prevent that again, but exporting the real declaration will allow you to write `prefix.ClassName.staticMethod(1)`, which the `Type` does not.

(Even with the "pretend Type objects have static methods, but only when used as literals" interpretation of static lookup, a getter invocation is not a type literal).

Reply:
Done.